[LWG motion 8] P2845R8 Formatting of std::filesystem::path (#6910)
cplusplus/draft: source/iostreams.tex

Equivalent to: \tcode{return path(lhs) /= rhs;}
\end{itemdescr}

\rSec3[fs.path.fmtr]{Formatter support}

\rSec4[fs.path.fmtr.general]{Formatter support overview}

\indexlibraryglobal{formatter}%
\begin{codeblock}
namespace std {
  template<class charT> struct formatter<filesystem::path, charT> {
    constexpr void set_debug_format();

    constexpr typename basic_format_parse_context<charT>::iterator
      parse(basic_format_parse_context<charT>& ctx);

    template<class FormatContext>
      typename FormatContext::iterator
        format(const filesystem::path& path, FormatContext& ctx) const;
  };
}
\end{codeblock}

> \begin{ncbnf}

Author: We don't typically add a \pnum before grammar rules (see [fs.path.generic]). That said, we usually introduce them with a paragraph, for example in [format.range.formatter]:

```
\pnum
\tcode{range_formatter} interprets \fmtgrammarterm{format-spec}
as a \fmtgrammarterm{range-format-spec}.
The syntax of format specifications is as follows:

\begin{ncbnf}
\fmtnontermdef{range-format-spec}\br
    \opt{range-fill-and-align} \opt{width} \opt{\terminal{n}} \opt{range-type} \opt{range-underlying-spec}
\end{ncbnf}

[...]
```

> \begin{ncbnf}

Reviewer: Right, you're just flowing straight from a class synopsis to a grammar rule to normal prose, with no pnum anywhere. The incoming paper had this grammar rule as part of a paragraph, but you've chosen to move it earlier. That seems reasonable, but you still need a pnum somewhere.

> \begin{ncbnf}

Reviewer 2: I added an introductory half-sentence to make this a proper paragraph.

the \fmtgrammarterm{path-format-spec} parsed by the last call to \tcode{parse}
contained the \tcode{?} option.
\end{itemdescr}

\indexlibrarymember{formatter}{basic_format_parse_context}%
\begin{itemdecl}
constexpr typename basic_format_parse_context<charT>::iterator
  parse(basic_format_parse_context<charT>& ctx);
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Parses the format specifier as a \fmtgrammarterm{path-format-spec} and
stores the parsed specifiers in \tcode{*this}.

\pnum
\returns
An iterator past the end of the \fmtgrammarterm{path-format-spec}.
\end{itemdescr}

\indexlibrarymember{formatter}{format}%
\begin{itemdecl}
template<class FormatContext>
  typename FormatContext::iterator
    format(const filesystem::path& p, FormatContext& ctx) const;
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Let \tcode{s} be \tcode{p.generic_string<filesystem::path::value_type>()}
if the \tcode{g} option is used,
otherwise \tcode{p.native()}.
Writes \tcode{s} into \tcode{ctx.out()},
adjusted according to the \fmtgrammarterm{path-format-spec}.
If \tcode{charT} is \keyword{char},
\tcode{path::value_type} is \keyword{wchar_t}, and
the literal encoding is UTF-8,

> then the escaped path is transcoded from the native encoding for

Reviewer: I don't think "the escaped path" is correct here, but that's what the paper says so I'll ask LWG.

Reviewer: This needs an LWG issue, the paper is wrong.

Reviewer 2: Thanks! I'll leave this with you if that's OK?

Reviewer: This is now [LWG 4070](https://cplusplus.github.io/LWG/issue4070).

wide character strings to UTF-8 with
maximal subparts of ill-formed subsequences
substituted with \ucode{fffd} \uname{replacement character}
per the Unicode Standard, Chapter 3.9 \ucode{fffd} Substitution in Conversion.
If \tcode{charT} and \tcode{path::value_type} are the same
then no transcoding is performed.
Otherwise, transcoding is

> \impldef{transcoding of the escaped path when \tcode{charT} and \tcode{path::value_type} differ}.

Reviewer: This is wrong. When the character types differ, transcoding is required whether the path is escaped or not.

```suggestion
\impldef{transcoding of a formatted path when \tcode{charT} and \tcode{path::value_type} differ}.
```

Reviewer: We _could_ address this as part of the LWG I'm going to open, but I worry that the index text will be forgotten, because it's not visible in this subclause (only the words "implementation defined" are shown here, the rest is in the annex).

Author: please open an issue?

Reviewer: We don't open issues against unmerged wording.

Reviewer 2: I'll use your suggestion for now, and I'll send you a reminder when you've opened the issue.

Reviewer: This is covered by [LWG 4070](https://cplusplus.github.io/LWG/issue4070) too.
